### PR TITLE
clamping sync requests

### PR DESCRIPTION
**Pull request by smnzhu**
closes #https://github.com/dapperlabs/flow-go/issues/5836

### CODE REVIEW

File: engine/common/synchronization/request_handler.go [hunk 1]
```diff
@@ -221,6 +222,11 @@ func (r *RequestHandlerEngine) onRangeRequest(originID flow.Identifier, req *mes
 		return nil
 	}
 
+	maxHeight := req.FromHeight + uint64(r.core.(*synchronization.Core).Config.MaxSize)
+	if maxHeight < req.FromHeight {
+		req.FromHeight = maxHeight
+	}
```
Comment (huitseeker) [inline engine/common/synchronization/request_handler.go:228]:
```suggestion
	if maxHeight < req.ToHeight {
		req.ToHeight = maxHeight
	}
```

File: engine/common/synchronization/request_handler.go [hunk 2]
```diff
@@ -266,7 +272,10 @@ func (r *RequestHandlerEngine) onBatchRequest(originID flow.Identifier, req *mes
 
 	// deduplicate the block IDs in the batch request
 	blockIDs := make(map[flow.Identifier]struct{})
-	for _, blockID := range req.BlockIDs {
+	for i, blockID := range req.BlockIDs {
+		if i == int(r.core.(*synchronization.Core).Config.MaxSize) {
+			break
+		}
```
Comment (huitseeker) [inline engine/common/synchronization/request_handler.go:278]: Is a client susceptible to sending a request for a small amount of BlockIDs that is so full of duplicates that it appears to be `> r.core.(*synchronization.Core).Config.MaxSize`?

Could we instead check `len(blockIDs)` as an additional condition to entering the loop?

**Review by huitseeker — COMMENTED** [review]
Thanks a bunch for getting to this fast!

I would love a code comment on each of the two checks making it explicit we enforce *existing*  client-side limits, and that this therefore shouldn't have an effect on a well-behaving client. 

Without that, another later reader, may not find it obvious those lines are meant to check for mischief.

**Comment by huitseeker** [comment]
One more thing:
`TestOnRangeRequest` and `TestOnBatchRequest` look ripe for a test exercising the over-the-limit requests.

**Comment by smnzhu** [comment]
Addressed comments

File: engine/common/synchronization/engine_test.go [hunk 3]
```diff
@@ -322,6 +322,21 @@ func (ss *SyncSuite) TestOnRangeRequest() {
...
+		func(args mock.Arguments) {
+			res := args.Get(0).(*messages.BlockResponse)
+			assert.Len(ss.T(), res.Blocks, int(synccore.DefaultConfig().MaxSize))
+			assert.Equal(ss.T(), req.Nonce, res.Nonce, "response should contain request nonce")
+			recipientID := args.Get(1).(flow.Identifier)
+			assert.Equal(ss.T(), originID, recipientID, "should send response to original requester")
+		},
+	)
+	err = ss.e.requestHandler.onRangeRequest(originID, req)
+	require.NoError(ss.T(), err, "valid range request should pass")
```
Comment (huitseeker) [inline engine/common/synchronization/engine_test.go:339]: Nit: do we want to call this "valid"?

File: engine/common/synchronization/engine_test.go [hunk 4]
```diff
@@ -358,9 +373,28 @@ func (ss *SyncSuite) TestOnBatchRequest() {
...
+		func(args mock.Arguments) {
+			res := args.Get(0).(*messages.BlockResponse)
+			assert.Len(ss.T(), res.Blocks, int(synccore.DefaultConfig().MaxSize))
+			assert.Equal(ss.T(), req.Nonce, res.Nonce, "response should contain request nonce")
+			recipientID := args.Get(1).(flow.Identifier)
+			assert.Equal(ss.T(), originID, recipientID, "response should be send to original requester")
+		},
+	)
+	err = ss.e.requestHandler.onBatchRequest(originID, req)
+	require.NoError(ss.T(), err, "should pass request with valid blocks")
```
Comment (huitseeker) [inline engine/common/synchronization/engine_test.go:397]: Nit: do we want to call this "valid"?

**Review by huitseeker — APPROVED** [review]
Thanks a bunch!

**Review by vishalchangrani — APPROVED** [review]
lgtm

**Comment by smnzhu** [comment]
bors merge

**Pull request merged**